packages/gatsby/src/redux/actions.js

```diff
@@ -366,6 +366,11 @@ actions.createNode = (node: any, plugin?: Plugin, traceId?: string) => {
     )
   }
 
+  // Make deep copy of node before mutating it, this makes sure original object
+  // will not fail validation due to unexpected fields.
+  // (reference: https://github.com/gatsbyjs/gatsby/issues/2772 )
+  node = _.cloneDeep(node)
```

Comment:
Sorry! this isn't what I meant. I'd like to avoid a deep clone entirely because it can very expensive, and we have no idea what content or structure nodes will have. Instead I'd like to find the places where mutation is happening and remove them, rather then make mutation safe.

Reply:
oh, right, I had brain lag or something - now i read your comment again and think to myself how did I understand it the way I did before ...

So if gatsby shouldn't mutate then you would have revert this - https://github.com/gatsbyjs/gatsby/commit/656299d2244ecbda35458cc262aea1cf146fca78

Reply:
Hmmmm yeah we could move that code into a library that then is used in source/transformer plugins that need it.

Reply:
yeah this is a bit tough, since we sort of need to hope consumers play nice and don't mutate...the problem tho is that enforcing generally it is pretty costly especially here where nodes may be/are complex data structures.

Reply:
Could you elaborate a little what problem https://github.com/gatsbyjs/gatsby/commit/656299d2244ecbda35458cc262aea1cf146fca78 fixed? Is it needed in all nodes or is just for source/transformers plugins? Maybe I could just change createNode and add argument to opt-out of adding `_PARENT` fields and just opt-out of it for plugin data? This would avoid any cloning at all (obviously prefered solution).

Reply:
See this comment https://github.com/gatsbyjs/gatsby/issues/2567#issuecomment-340068410 It's basically a workaround for GraphQL as it doesn't let you know what the parent of a sub-object is which breaks Gatsby in subtle ways.

Avoiding mutating plugins would be a great start. So yeah, question is should we opt-in to adding _PARENT or opt-out.

I think it's a generally useful/needed pattern which is why I added it to actions.js so I'm leaning towards opt-out. Thoughts?

Reply:
maybe instead of tagging each sub object we could use a WeakMap hashing objects to parent id? e.g. something like:

```
map = new WeakMap()

eachChildObject(node, obj => { 
  map.set(obj, node.id)
})
```

then expose a getParent api perhaps?

Reply:
Ooo that's good! Yeah, that sounds much nicer. @pieh you have time to work on this by chance? I'm pretty strapped for time over next few days / week.